{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "18682e0e-4c0b-40cd-b415-350dca147388",
   "metadata": {},
   "outputs": [],
   "source": [
    "mamba activate macs3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4deb4005-e5d0-49a2-b691-c43cae0e0086",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: macs3 [-h] [--version]\n",
      "             {callpeak,bdgpeakcall,bdgbroadcall,bdgcmp,bdgopt,cmbreps,bdgdiff,filterdup,predictd,pileup,randsample,refinepeak,callvar,hmmratac}\n",
      "             ...\n",
      "\n",
      "macs3 -- Model-based Analysis for ChIP-Sequencing\n",
      "\n",
      "positional arguments:\n",
      "  {callpeak,bdgpeakcall,bdgbroadcall,bdgcmp,bdgopt,cmbreps,bdgdiff,filterdup,predictd,pileup,randsample,refinepeak,callvar,hmmratac}\n",
      "    callpeak            Main MACS3 Function: Call peaks from alignment\n",
      "                        results.\n",
      "    bdgpeakcall         Call peaks from bedGraph file.\n",
      "    bdgbroadcall        Call nested broad peaks from bedGraph file.\n",
      "    bdgcmp              Comparing two signal tracks in bedGraph format.\n",
      "    bdgopt              Operate the score column of bedGraph file.\n",
      "    cmbreps             Combine bedGraph files of scores from replicates.\n",
      "    bdgdiff             Differential peak detection based on paired four\n",
      "                        bedGraph files.\n",
      "    filterdup           Remove duplicate reads, then save in BED/BEDPE format\n",
      "                        file.\n",
      "    predictd            Predict d or fragment size from alignment results. In\n",
      "                        case of PE data, report the average insertion/fragment\n",
      "                        size from all pairs.\n",
      "    pileup              Pileup aligned reads (single-end) or fragments\n",
      "                        (paired-end).\n",
      "    randsample          Randomly choose a number/percentage of total reads,\n",
      "                        then save in BED/BEDPE format file.\n",
      "    refinepeak          Take raw reads alignment, refine peak summits.\n",
      "                        Inspired by SPP.\n",
      "    callvar             Call variants in given peak regions from the alignment\n",
      "                        BAM files.\n",
      "    hmmratac            Dedicated peak calling based on Hidden Markov Model\n",
      "                        for ATAC-seq data.\n",
      "\n",
      "options:\n",
      "  -h, --help            show this help message and exit\n",
      "  --version             show program's version number and exit\n",
      "\n",
      "For command line options of each command, type: macs3 COMMAND -h\n"
     ]
    }
   ],
   "source": [
    "macs3 -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "40386bab-53e4-4d15-b47a-f83a37a980e6",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: macs3 callpeak [-h] -t TFILE [TFILE ...] [-c [CFILE ...]]\n",
      "                      [-f {AUTO,BAM,SAM,BED,ELAND,ELANDMULTI,ELANDEXPORT,BOWTIE,BAMPE,BEDPE}]\n",
      "                      [-g GSIZE] [-s TSIZE] [--keep-dup KEEPDUPLICATES]\n",
      "                      [--outdir OUTDIR] [-n NAME] [-B] [--verbose VERBOSE]\n",
      "                      [--trackline] [--SPMR] [--nomodel] [--shift SHIFT]\n",
      "                      [--extsize EXTSIZE] [--bw BW] [--d-min D_MIN]\n",
      "                      [-m MFOLD MFOLD] [--fix-bimodal] [-q QVALUE | -p PVALUE]\n",
      "                      [--scale-to {large,small}] [--down-sample] [--seed SEED]\n",
      "                      [--tempdir TEMPDIR] [--nolambda] [--slocal SMALLLOCAL]\n",
      "                      [--llocal LARGELOCAL] [--max-gap MAXGAP]\n",
      "                      [--min-length MINLEN] [--broad]\n",
      "                      [--broad-cutoff BROADCUTOFF] [--cutoff-analysis]\n",
      "                      [--call-summits] [--fe-cutoff FECUTOFF] [--to-large]\n",
      "                      [--ratio RATIO] [--buffer-size BUFFER_SIZE]\n",
      "\n",
      "options:\n",
      "  -h, --help            show this help message and exit\n",
      "\n",
      "Input files arguments:\n",
      "  -t TFILE [TFILE ...], --treatment TFILE [TFILE ...]\n",
      "                        ChIP-seq treatment file. If multiple files are given\n",
      "                        as '-t A B C', then they will all be read and pooled\n",
      "                        together. REQUIRED.\n",
      "  -c [CFILE ...], --control [CFILE ...]\n",
      "                        Control file. If multiple files are given as '-c A B\n",
      "                        C', they will be pooled to estimate ChIP-seq\n",
      "                        background noise.\n",
      "  -f {AUTO,BAM,SAM,BED,ELAND,ELANDMULTI,ELANDEXPORT,BOWTIE,BAMPE,BEDPE}, --format {AUTO,BAM,SAM,BED,ELAND,ELANDMULTI,ELANDEXPORT,BOWTIE,BAMPE,BEDPE}\n",
      "                        Format of tag file, \"AUTO\", \"BED\" or \"ELAND\" or\n",
      "                        \"ELANDMULTI\" or \"ELANDEXPORT\" or \"SAM\" or \"BAM\" or\n",
      "                        \"BOWTIE\" or \"BAMPE\" or \"BEDPE\". The default AUTO\n",
      "                        option will let MACS decide which format (except for\n",
      "                        BAMPE and BEDPE which should be implicitly set) the\n",
      "                        file is. Please check the definition in README. Please\n",
      "                        note that if the format is set as BAMPE or BEDPE,\n",
      "                        MACS3 will call its special Paired-end mode to call\n",
      "                        peaks by piling up the actual ChIPed fragments defined\n",
      "                        by both aligned ends, instead of predicting the\n",
      "                        fragment size first and extending reads. Also please\n",
      "                        note that the BEDPE only contains three columns, and\n",
      "                        is NOT the same BEDPE format used by BEDTOOLS.\n",
      "                        DEFAULT: \"AUTO\"\n",
      "  -g GSIZE, --gsize GSIZE\n",
      "                        Effective genome size. It can be 1.0e+9 or 1000000000,\n",
      "                        or shortcuts:'hs' for human (2,913,022,398), 'mm' for\n",
      "                        mouse (2,652,783,500), 'ce' for C. elegans\n",
      "                        (100,286,401) and 'dm' for fruitfly (142,573,017),\n",
      "                        Default:hs. The effective genome size numbers for the\n",
      "                        above four species are collected from Deeptools https:\n",
      "                        //deeptools.readthedocs.io/en/develop/content/feature/\n",
      "                        effectiveGenomeSize.html Please refer to deeptools to\n",
      "                        define the best genome size you plan to use.\n",
      "  -s TSIZE, --tsize TSIZE\n",
      "                        Tag size/read length. This will override the auto\n",
      "                        detected tag size. DEFAULT: Not set\n",
      "  --keep-dup KEEPDUPLICATES\n",
      "                        It controls the behavior towards duplicate tags at the\n",
      "                        exact same location -- the same coordination and the\n",
      "                        same strand. The 'auto' option makes MACS calculate\n",
      "                        the maximum tags at the exact same location based on\n",
      "                        binomal distribution using 1e-5 as pvalue cutoff; and\n",
      "                        the 'all' option keeps every tags. If an integer is\n",
      "                        given, at most this number of tags will be kept at the\n",
      "                        same location. Note, if you've used samtools or picard\n",
      "                        to flag reads as 'PCR/Optical duplicate' in bit 1024,\n",
      "                        MACS3 will still read them although the reads may be\n",
      "                        decided by MACS3 as duplicate later. If you plan to\n",
      "                        rely on samtools/picard/any other tool to filter\n",
      "                        duplicates, please remove those duplicate reads and\n",
      "                        save a new alignment file then ask MACS3 to keep all\n",
      "                        by '--keep-dup all'. The default is to keep one tag at\n",
      "                        the same location. Default: 1\n",
      "\n",
      "Output arguments:\n",
      "  --outdir OUTDIR       If specified all output files will be written to that\n",
      "                        directory. Default: the current working directory\n",
      "  -n NAME, --name NAME  Experiment name, which will be used to generate output\n",
      "                        file names. DEFAULT: \"NA\"\n",
      "  -B, --bdg             Whether or not to save extended fragment pileup, and\n",
      "                        local lambda tracks (two files) at every bp into a\n",
      "                        bedGraph file. DEFAULT: False\n",
      "  --verbose VERBOSE     Set verbose level of runtime message. 0: only show\n",
      "                        critical message, 1: show additional warning message,\n",
      "                        2: show process information, 3: show debug messages.\n",
      "                        DEFAULT:2\n",
      "  --trackline           Tells MACS to include trackline with bedGraph files.\n",
      "                        To include this trackline while displaying bedGraph at\n",
      "                        UCSC genome browser, can show name and description of\n",
      "                        the file as well. However my suggestion is to convert\n",
      "                        bedGraph to bigWig, then show the smaller and faster\n",
      "                        binary bigWig file at UCSC genome browser, as well as\n",
      "                        downstream analysis. Require -B to be set. Default:\n",
      "                        Not include trackline.\n",
      "  --SPMR                If True, MACS will SAVE signal per million reads for\n",
      "                        fragment pileup profiles. It won't interfere with\n",
      "                        computing pvalue/qvalue during peak calling, since\n",
      "                        internally MACS3 keeps using the raw pileup and\n",
      "                        scaling factors between larger and smaller dataset to\n",
      "                        calculate statistics measurements. If you plan to use\n",
      "                        the signal output in bedGraph to call peaks using\n",
      "                        bdgcmp and bdgpeakcall, you shouldn't use this option\n",
      "                        because you will end up with different results.\n",
      "                        However, this option is recommended for displaying\n",
      "                        normalized pileup tracks across many datasets. Require\n",
      "                        -B to be set. Default: False\n",
      "\n",
      "Shifting model arguments:\n",
      "  --nomodel             Whether or not to build the shifting model. If True,\n",
      "                        MACS will not build model. by default it means\n",
      "                        shifting size = 100, try to set extsize to change it.\n",
      "                        It's highly recommended that while you have many\n",
      "                        datasets to process and you plan to compare different\n",
      "                        conditions, aka differential calling, use both\n",
      "                        'nomodel' and 'extsize' to make signal files from\n",
      "                        different datasets comparable. DEFAULT: False\n",
      "  --shift SHIFT         (NOT the legacy --shiftsize option!) The arbitrary\n",
      "                        shift in bp. Use discretion while setting it other\n",
      "                        than default value. When NOMODEL is set, MACS will use\n",
      "                        this value to move cutting ends (5') towards 5'->3'\n",
      "                        direction then apply EXTSIZE to extend them to\n",
      "                        fragments. When this value is negative, ends will be\n",
      "                        moved toward 3'->5' direction. Recommended to keep it\n",
      "                        as default 0 for ChIP-Seq datasets, or -1 * half of\n",
      "                        EXTSIZE together with EXTSIZE option for detecting\n",
      "                        enriched cutting loci such as certain DNAseI-Seq\n",
      "                        datasets. Note, you can't set values other than 0 if\n",
      "                        format is BAMPE or BEDPE for paired-end data. DEFAULT:\n",
      "                        0.\n",
      "  --extsize EXTSIZE     The arbitrary extension size in bp. When nomodel is\n",
      "                        true, MACS will use this value as fragment size to\n",
      "                        extend each read towards 3' end, then pile them up.\n",
      "                        It's exactly twice the number of obsolete SHIFTSIZE.\n",
      "                        In previous language, each read is moved 5'->3'\n",
      "                        direction to middle of fragment by 1/2 d, then\n",
      "                        extended to both direction with 1/2 d. This is\n",
      "                        equivalent to say each read is extended towards 5'->3'\n",
      "                        into a d size fragment. DEFAULT: 200. EXTSIZE and\n",
      "                        SHIFT can be combined when necessary. Check SHIFT\n",
      "                        option.\n",
      "  --bw BW               Band width for picking regions to compute fragment\n",
      "                        size. This value is only used while building the\n",
      "                        shifting model. Tweaking this is not recommended.\n",
      "                        DEFAULT: 300\n",
      "  --d-min D_MIN         Minimum fragment size in basepair. Any predicted\n",
      "                        fragment size less than this will be excluded.\n",
      "                        DEFAULT: 20\n",
      "  -m MFOLD MFOLD, --mfold MFOLD MFOLD\n",
      "                        Select the regions within MFOLD range of high-\n",
      "                        confidence enrichment ratio against background to\n",
      "                        build model. Fold-enrichment in regions must be lower\n",
      "                        than upper limit, and higher than the lower limit. Use\n",
      "                        as \"-m 10 30\". This setting is only used while\n",
      "                        building the shifting model. Tweaking it is not\n",
      "                        recommended. DEFAULT:5 50\n",
      "  --fix-bimodal         Whether turn on the auto pair model process. If set,\n",
      "                        when MACS failed to build paired model, it will use\n",
      "                        the nomodel settings, the --exsize parameter to extend\n",
      "                        each tags towards 3' direction. Not to use this\n",
      "                        automate fixation is a default behavior now. DEFAULT:\n",
      "                        False\n",
      "\n",
      "Peak calling arguments:\n",
      "  -q QVALUE, --qvalue QVALUE\n",
      "                        Minimum FDR (q-value) cutoff for peak detection.\n",
      "                        DEFAULT: 0.05. -q, and -p are mutually exclusive.\n",
      "  -p PVALUE, --pvalue PVALUE\n",
      "                        Pvalue cutoff for peak detection. DEFAULT: not set.\n",
      "                        -q, and -p are mutually exclusive. If pvalue cutoff is\n",
      "                        set, qvalue will not be calculated and reported as -1\n",
      "                        in the final .xls file.\n",
      "  --scale-to {large,small}\n",
      "                        When set to 'small', scale the larger sample up to the\n",
      "                        smaller sample. When set to 'larger', scale the\n",
      "                        smaller sample up to the bigger sample. By default,\n",
      "                        scale to 'small'. This option replaces the obsolete '\n",
      "                        --to-large' option. The default behavior is\n",
      "                        recommended since it will lead to less significant\n",
      "                        p/q-values in general but more specific results. Keep\n",
      "                        in mind that scaling down will influence control/input\n",
      "                        sample more. DEFAULT: 'small', the choice is either\n",
      "                        'small' or 'large'.\n",
      "  --down-sample         When set, random sampling method will scale down the\n",
      "                        bigger sample. By default, MACS uses linear scaling.\n",
      "                        Warning: This option will make your result unstable\n",
      "                        and irreproducible since each time, random reads would\n",
      "                        be selected. Consider to use 'randsample' script\n",
      "                        instead. <not implmented>If used together with --SPMR,\n",
      "                        1 million unique reads will be randomly picked.</not\n",
      "                        implemented> Caution: due to the implementation, the\n",
      "                        final number of selected reads may not be as you\n",
      "                        expected! DEFAULT: False\n",
      "  --seed SEED           Set the random seed while down sampling data. Must be\n",
      "                        a non-negative integer in order to be effective.\n",
      "                        DEFAULT: not set\n",
      "  --tempdir TEMPDIR     Optional directory to store temp files. DEFAULT:\n",
      "                        /scratch/tmp.56223913.jjanssens\n",
      "  --nolambda            If True, MACS will use fixed background lambda as\n",
      "                        local lambda for every peak region. Normally, MACS\n",
      "                        calculates a dynamic local lambda to reflect the local\n",
      "                        bias due to the potential chromatin accessibility.\n",
      "  --slocal SMALLLOCAL   The small nearby region in basepairs to calculate\n",
      "                        dynamic lambda. This is used to capture the bias near\n",
      "                        the peak summit region. Invalid if there is no control\n",
      "                        data. If you set this to 0, MACS will skip slocal\n",
      "                        lambda calculation. *Note* that MACS will always\n",
      "                        perform a d-size local lambda calculation while the\n",
      "                        control data is available. The final local bias would\n",
      "                        be the maximum of the lambda value from d, slocal, and\n",
      "                        llocal size windows. While control is not available, d\n",
      "                        and slocal lambda won't be considered. DEFAULT: 1000\n",
      "  --llocal LARGELOCAL   The large nearby region in basepairs to calculate\n",
      "                        dynamic lambda. This is used to capture the surround\n",
      "                        bias. If you set this to 0, MACS will skip llocal\n",
      "                        lambda calculation. *Note* that MACS will always\n",
      "                        perform a d-size local lambda calculation while the\n",
      "                        control data is available. The final local bias would\n",
      "                        be the maximum of the lambda value from d, slocal, and\n",
      "                        llocal size windows. While control is not available, d\n",
      "                        and slocal lambda won't be considered. DEFAULT: 10000.\n",
      "  --max-gap MAXGAP      Maximum gap between significant sites to cluster them\n",
      "                        together. The DEFAULT value is the detected read\n",
      "                        length/tag size.\n",
      "  --min-length MINLEN   Minimum length of a peak. The DEFAULT value is the\n",
      "                        predicted fragment size d. Note, if you set a value\n",
      "                        smaller than the fragment size, it may have NO effect\n",
      "                        on the result. For BROAD peak calling, try to set a\n",
      "                        large value such as 500bps. You can also use '--\n",
      "                        cutoff-analysis' option with default setting, and\n",
      "                        check the column 'avelpeak' under different cutoff\n",
      "                        values to decide a reasonable minlen value.\n",
      "  --broad               If set, MACS will try to call broad peaks using the\n",
      "                        --broad-cutoff setting. Please tweak '--broad-cutoff'\n",
      "                        setting to control the peak calling behavior. At the\n",
      "                        meantime, either -q or -p cutoff will be used to\n",
      "                        define regions with 'stronger enrichment' inside of\n",
      "                        broad peaks. The maximum gap is expanded to 4 * MAXGAP\n",
      "                        (--max-gap parameter). As a result, MACS will output a\n",
      "                        'gappedPeak' and a 'broadPeak' file instead of\n",
      "                        'narrowPeak' file. Note, a broad peak will be reported\n",
      "                        even if there is no 'stronger enrichment' inside.\n",
      "                        DEFAULT: False\n",
      "  --broad-cutoff BROADCUTOFF\n",
      "                        Cutoff for broad region. This option is not available\n",
      "                        unless --broad is set. If -p is set, this is a pvalue\n",
      "                        cutoff, otherwise, it's a qvalue cutoff. Please note\n",
      "                        that in broad peakcalling mode, MACS3 uses this\n",
      "                        setting to control the overall peak calling behavior,\n",
      "                        then uses -q or -p setting to define regions inside\n",
      "                        broad region as 'stronger' enrichment. DEFAULT: 0.1\n",
      "  --cutoff-analysis     While set, MACS3 will analyze number or total length\n",
      "                        of peaks that can be called by different p-value\n",
      "                        cutoff then output a summary table to help user decide\n",
      "                        a better cutoff. The table will be saved in\n",
      "                        NAME_cutoff_analysis.txt file. Note, minlen and maxgap\n",
      "                        may affect the results. WARNING: May take ~30 folds\n",
      "                        longer time to finish. The result can be useful for\n",
      "                        users to decide a reasonable cutoff value. DEFAULT:\n",
      "                        False\n",
      "\n",
      "Post-processing options:\n",
      "  --call-summits        If set, MACS will use a more sophisticated signal\n",
      "                        processing approach to find subpeak summits in each\n",
      "                        enriched peak region. DEFAULT: False\n",
      "  --fe-cutoff FECUTOFF  When set, the value will be used as the minimum\n",
      "                        requirement to filter out peaks with low fold-\n",
      "                        enrichment. Note, MACS3 adds one as pseudocount while\n",
      "                        calculating fold-enrichment. By default, it is set as\n",
      "                        1 so there is no filtering. DEFAULT: 1.0\n",
      "\n",
      "Obsolete options:\n",
      "  --to-large            Obsolete option. Please use '--scale-to large'\n",
      "                        instead.\n",
      "  --ratio RATIO         Obsolete option. Originally designed to normalize\n",
      "                        treatment and control with customized ratio, now it\n",
      "                        won't have any effect.\n",
      "\n",
      "Other options:\n",
      "  --buffer-size BUFFER_SIZE\n",
      "                        Buffer size for incrementally increasing internal\n",
      "                        array size to store reads alignment information. In\n",
      "                        most cases, you don't have to change this parameter.\n",
      "                        However, if there are large number of\n",
      "                        chromosomes/contigs/scaffolds in your alignment, it's\n",
      "                        recommended to specify a smaller buffer size in order\n",
      "                        to decrease memory usage (but it will take longer time\n",
      "                        to read alignment files). Minimum memory requested for\n",
      "                        reading an alignment file is about # of CHROMOSOME *\n",
      "                        BUFFER_SIZE * 8 Bytes. DEFAULT: 100000\n",
      "\n",
      "Examples:\n",
      "1. Peak calling for regular TF ChIP-seq:\n",
      "    $ macs3 callpeak -t ChIP.bam -c Control.bam -f BAM -g hs -n test -B -q 0.01\n",
      "2. Broad peak calling on Histone Mark ChIP-seq:\n",
      "    $ macs3 callpeak -t ChIP.bam -c Control.bam --broad -g hs --broad-cutoff 0.1\n",
      "3. Peak calling on ATAC-seq (paired-end mode):\n",
      "    $ macs3 callpeak -f BAMPE -t ATAC.bam -g hs -n test -B -q 0.01\n",
      "4. Peak calling on ATAC-seq ( focusing on insertion sites, and using single-end mode):\n",
      "    $ macs3 callpeak -f BAM -t ATAC.bam -g hs -n test -B -q 0.01 --shift -50 --extension 100\n",
      "(macs3) \n"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "macs3 callpeak -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3b9f5bcd-de16-41fb-8b42-783bc17d19b6",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO  @ 24 Apr 2024 15:21:46: [30 MB] \n",
      "# Command line: callpeak -f BAMPE -t enterocytes_species/enterocyte_human_sorted_filtered.bam -g hs -n human_peaks -B -q 0.01\n",
      "# ARGUMENTS LIST:\n",
      "# name = human_peaks\n",
      "# format = BAMPE\n",
      "# ChIP-seq file = ['enterocytes_species/enterocyte_human_sorted_filtered.bam']\n",
      "# control file = None\n",
      "# effective genome size = 2.91e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 1.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is on\n",
      " \n",
      "INFO  @ 24 Apr 2024 15:21:46: [30 MB] #1 read fragment files... \n",
      "INFO  @ 24 Apr 2024 15:21:46: [30 MB] #1 read treatment fragments... \n",
      "INFO  @ 24 Apr 2024 15:21:50: [66 MB]  1000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:21:53: [74 MB]  2000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:21:57: [103 MB]  3000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:22:01: [103 MB]  4000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:22:04: [103 MB]  5000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:22:08: [105 MB]  6000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:22:12: [113 MB]  7000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:22:15: [113 MB]  8000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:22:19: [120 MB]  9000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:22:22: [128 MB]  10000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:22:26: [136 MB]  11000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:22:30: [144 MB]  12000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:22:33: [152 MB]  13000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:22:37: [159 MB]  14000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:22:40: [167 MB]  15000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:22:44: [175 MB]  16000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:22:47: [182 MB]  17000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:22:51: [191 MB]  18000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:22:55: [198 MB]  19000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:22:58: [206 MB]  20000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:23:02: [213 MB]  21000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:23:05: [221 MB]  22000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:23:09: [229 MB]  23000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:23:12: [236 MB]  24000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:23:16: [244 MB]  25000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:23:20: [254 MB]  26000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:23:23: [261 MB]  27000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:23:27: [269 MB]  28000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:23:30: [277 MB]  29000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:23:34: [285 MB]  30000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:23:37: [293 MB]  31000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:23:41: [301 MB]  32000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:23:44: [308 MB]  33000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:23:48: [316 MB]  34000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:23:52: [325 MB]  35000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:23:55: [332 MB]  36000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:23:59: [340 MB]  37000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:24:02: [348 MB]  38000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:24:06: [355 MB]  39000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:24:09: [364 MB]  40000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:24:13: [372 MB]  41000000 fragments parsed \n",
      "INFO  @ 24 Apr 2024 15:24:15: [432 MB] 41894103 fragments have been read. \n",
      "INFO  @ 24 Apr 2024 15:24:45: [432 MB] #1 mean fragment size is determined as 164.4 bp from treatment \n",
      "INFO  @ 24 Apr 2024 15:24:45: [432 MB] #1 fragment size = 164.4 \n",
      "INFO  @ 24 Apr 2024 15:24:45: [432 MB] #1  total fragments in treatment: 41894103 \n",
      "INFO  @ 24 Apr 2024 15:24:45: [432 MB] #1 user defined the maximum fragments... \n",
      "INFO  @ 24 Apr 2024 15:24:45: [432 MB] #1 filter out redundant fragments by allowing at most 1 identical fragment(s) \n",
      "INFO  @ 24 Apr 2024 15:25:39: [485 MB] #1  fragments after filtering in treatment: 40498977 \n",
      "INFO  @ 24 Apr 2024 15:25:39: [485 MB] #1  Redundant rate of treatment: 0.03 \n",
      "INFO  @ 24 Apr 2024 15:25:39: [485 MB] #1 finished! \n",
      "INFO  @ 24 Apr 2024 15:25:39: [485 MB] #2 Build Peak Model... \n",
      "INFO  @ 24 Apr 2024 15:25:39: [485 MB] #2 Skipped... \n",
      "INFO  @ 24 Apr 2024 15:25:39: [485 MB] #3 Call peaks... \n",
      "INFO  @ 24 Apr 2024 15:25:39: [485 MB] #3 Pre-compute pvalue-qvalue table... \n",
      "INFO  @ 24 Apr 2024 15:27:52: [1477 MB] #3 In the peak calling step, the following will be performed simultaneously: \n",
      "INFO  @ 24 Apr 2024 15:27:52: [1477 MB] #3   Write bedGraph files for treatment pileup (after scaling if necessary)... human_peaks_treat_pileup.bdg \n",
      "INFO  @ 24 Apr 2024 15:27:52: [1477 MB] #3   Write bedGraph files for control lambda (after scaling if necessary)... human_peaks_control_lambda.bdg \n",
      "INFO  @ 24 Apr 2024 15:27:52: [1477 MB] #3   Pileup will be based on sequencing depth in treatment. \n",
      "INFO  @ 24 Apr 2024 15:27:52: [1477 MB] #3 Call peaks for each chromosome... \n",
      "INFO  @ 24 Apr 2024 15:30:08: [1477 MB] #4 Write output xls file... human_peaks_peaks.xls \n",
      "INFO  @ 24 Apr 2024 15:30:08: [1477 MB] #4 Write peak in narrowPeak format file... human_peaks_peaks.narrowPeak \n",
      "INFO  @ 24 Apr 2024 15:30:08: [1477 MB] #4 Write summits bed file... human_peaks_summits.bed \n",
      "INFO  @ 24 Apr 2024 15:30:08: [1477 MB] Done! \n",
      "(macs3) \n"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "macs3 callpeak -f BAMPE -t enterocytes_species/enterocyte_human_sorted_filtered.bam -g hs -n called_peaks/human_peaks -B -q 0.01\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fc7ed7c1-9a1a-4d01-b15d-5f2bd020aba3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO  @ 12 Jun 2024 03:26:43: [39 MB] \n",
      "# Command line: callpeak -f BAMPE -t enterocytes_species/enterocyte_chimp_sorted_filtered.bam -g hs -n called_peaks/chimp_peaks -B -q 0.01\n",
      "# ARGUMENTS LIST:\n",
      "# name = called_peaks/chimp_peaks\n",
      "# format = BAMPE\n",
      "# ChIP-seq file = ['enterocytes_species/enterocyte_chimp_sorted_filtered.bam']\n",
      "# control file = None\n",
      "# effective genome size = 2.91e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 1.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is on\n",
      " \n",
      "INFO  @ 12 Jun 2024 03:26:43: [39 MB] #1 read fragment files... \n",
      "INFO  @ 12 Jun 2024 03:26:43: [39 MB] #1 read treatment fragments... \n",
      "INFO  @ 12 Jun 2024 03:26:46: [67 MB]  1000000 fragments parsed \n",
      "INFO  @ 12 Jun 2024 03:26:49: [75 MB]  2000000 fragments parsed \n",
      "INFO  @ 12 Jun 2024 03:26:52: [84 MB]  3000000 fragments parsed \n",
      "INFO  @ 12 Jun 2024 03:26:55: [93 MB]  4000000 fragments parsed \n",
      "INFO  @ 12 Jun 2024 03:26:58: [101 MB]  5000000 fragments parsed \n",
      "INFO  @ 12 Jun 2024 03:27:02: [109 MB]  6000000 fragments parsed \n",
      "INFO  @ 12 Jun 2024 03:27:05: [118 MB]  7000000 fragments parsed \n",
      "INFO  @ 12 Jun 2024 03:27:08: [126 MB]  8000000 fragments parsed \n",
      "INFO  @ 12 Jun 2024 03:27:11: [135 MB]  9000000 fragments parsed \n",
      "INFO  @ 12 Jun 2024 03:27:14: [143 MB]  10000000 fragments parsed \n",
      "INFO  @ 12 Jun 2024 03:27:17: [151 MB]  11000000 fragments parsed \n",
      "INFO  @ 12 Jun 2024 03:27:20: [161 MB]  12000000 fragments parsed \n",
      "INFO  @ 12 Jun 2024 03:27:23: [170 MB]  13000000 fragments parsed \n",
      "INFO  @ 12 Jun 2024 03:27:26: [178 MB]  14000000 fragments parsed \n",
      "INFO  @ 12 Jun 2024 03:27:29: [532 MB] 14769968 fragments have been read. \n",
      "INFO  @ 12 Jun 2024 03:27:38: [532 MB] #1 mean fragment size is determined as 166.3 bp from treatment \n",
      "INFO  @ 12 Jun 2024 03:27:38: [532 MB] #1 fragment size = 166.3 \n",
      "INFO  @ 12 Jun 2024 03:27:38: [532 MB] #1  total fragments in treatment: 14769968 \n",
      "INFO  @ 12 Jun 2024 03:27:38: [532 MB] #1 user defined the maximum fragments... \n",
      "INFO  @ 12 Jun 2024 03:27:38: [532 MB] #1 filter out redundant fragments by allowing at most 1 identical fragment(s) \n",
      "INFO  @ 12 Jun 2024 03:27:57: [545 MB] #1  fragments after filtering in treatment: 14742268 \n",
      "INFO  @ 12 Jun 2024 03:27:57: [545 MB] #1  Redundant rate of treatment: 0.00 \n",
      "INFO  @ 12 Jun 2024 03:27:57: [545 MB] #1 finished! \n",
      "INFO  @ 12 Jun 2024 03:27:57: [545 MB] #2 Build Peak Model... \n",
      "INFO  @ 12 Jun 2024 03:27:57: [545 MB] #2 Skipped... \n",
      "INFO  @ 12 Jun 2024 03:27:57: [545 MB] #3 Call peaks... \n",
      "INFO  @ 12 Jun 2024 03:27:57: [545 MB] #3 Pre-compute pvalue-qvalue table... \n",
      "INFO  @ 12 Jun 2024 03:28:37: [940 MB] #3 In the peak calling step, the following will be performed simultaneously: \n",
      "INFO  @ 12 Jun 2024 03:28:37: [940 MB] #3   Write bedGraph files for treatment pileup (after scaling if necessary)... called_peaks/chimp_peaks_treat_pileup.bdg \n",
      "INFO  @ 12 Jun 2024 03:28:37: [940 MB] #3   Write bedGraph files for control lambda (after scaling if necessary)... called_peaks/chimp_peaks_control_lambda.bdg \n",
      "INFO  @ 12 Jun 2024 03:28:37: [940 MB] #3   Pileup will be based on sequencing depth in treatment. \n",
      "INFO  @ 12 Jun 2024 03:28:37: [940 MB] #3 Call peaks for each chromosome... \n",
      "INFO  @ 12 Jun 2024 03:29:20: [940 MB] #4 Write output xls file... called_peaks/chimp_peaks_peaks.xls \n",
      "INFO  @ 12 Jun 2024 03:29:21: [940 MB] #4 Write peak in narrowPeak format file... called_peaks/chimp_peaks_peaks.narrowPeak \n",
      "INFO  @ 12 Jun 2024 03:29:21: [940 MB] #4 Write summits bed file... called_peaks/chimp_peaks_summits.bed \n",
      "INFO  @ 12 Jun 2024 03:29:21: [940 MB] Done! \n"
     ]
    }
   ],
   "source": [
    "macs3 callpeak -f BAMPE -t enterocytes_species/enterocyte_chimp_sorted_filtered.bam -g hs -n called_peaks/chimp_peaks -B -q 0.01\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "020e4071-8b1a-4a37-a33d-6e22f5ef4a4a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60846360-9311-40f7-bdd6-e779a4432a4b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "569fa862-579a-4da7-a415-663966f5f4f6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71dfbad5-7f22-4936-a821-2fe7fea64892",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25eeb9cc-4253-4f1d-91f4-2adebde9dcd5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
